File: 0872-leaf-similar-trees/0872-leaf-similar-trees.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def leafSimilar(self, root1, root2):
        """
        :type root1: TreeNode
        :type root2: TreeNode
        :rtype: bool
        """
        leaves1, leaves2 = [], []

        self.getLeafs(root1, leaves1)
        self.getLeafs(root2, leaves2)

        return leaves1 == leaves2

    def getLeafs(self, root, leaves):
        if not root:
            return
        
        if not root.left and not root.right:
            leaves.append(root.val)
        
        self.getLeafs(root.left, leaves)
        self.getLeafs(root.right, leaves)
        

File: 0872-leaf-similar-trees/0872-leaf-similar-trees.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func leafSimilar(root1 *TreeNode, root2 *TreeNode) bool {
    arr1, arr2 := []int{}, []int{}
    getLeafs(root1, &arr1)
    getLeafs(root2, &arr2)

    if len(arr1) != len(arr2) {
        return false
    }

    for i := range arr1 {
        if arr1[i] != arr2[i] {
            return false
        }
    }

    return true
}

func getLeafs(root *TreeNode, arr *[]int) {
    if root == nil {
        return 
    }

    if root.Left == nil && root.Right == nil {
        *arr = append(*arr, root.Val)
        return
    }

    getLeafs(root.Left, arr)
    getLeafs(root.Right, arr)
}